File: Practica/Guia08/src/huffman.py
#! /usr/bin/python3
# @author Juan Cruz Mateos
from pprint import pprint


class Nodo:
    def __init__(self, simb: str, prob: int, izq=None, der=None):
        self.simb = simb
        self.prob = prob
        self.izq = izq
        self.der = der

    def is_hoja(self):
        return self.izq == None and self.der == None

    def __str__(self) -> str:
        return f"simb={self.simb}, prob={self.prob}"


def calc_frecuencies(input: str) -> dict:
    frec = {}
    for char in input:
        frec[char] = frec.get(char, 0) + 1
    return frec


def calc_probabilities(frec: dict) -> dict:
    prob = {}
    total = sum(frec.values())
    for char in frec:
        prob[char] = frec[char] / total
    return prob


def create_tree(prob: dict) -> Nodo:
    minHeap = []
    for char, p in prob.items():
        minHeap.append(Nodo(char, p))

    while len(minHeap) > 1:
        minHeap.sort(key=lambda nodo: nodo.prob)
        # print(*minHeap)
        hijoIzq = minHeap.pop(0)
        hijoDer = minHeap.pop(0)
        # print(hijoIzq.simb, hijoIzq.prob, hijoDer.simb, hijoDer.prob)
        padre = Nodo(None, hijoIzq.prob + hijoDer.prob, hijoIzq, hijoDer)
        minHeap.append(padre)
    return minHeap.pop()


def encode(code: dict, nodo: Nodo, huff: str = ""):
    if nodo.is_hoja():
        code[nodo.simb] = huff
    else:
        encode(code, nodo.izq, huff + "0")
        encode(code, nodo.der, huff + "1")


def printable(c: str) -> str:
    if not c.isalpha():
        return "' '"
    return c


def print_codes(huff_codes: dict):
    print(f"{'simb':^10s}|{'huffman':^10s}")
    print('-' * 21)
    for simb, code in huff_codes.items():
        print(f"{printable(simb):^10s}|{code:^10s}")


def main():
    huff_codes = {}
    string = "CALIDO AMANECER SOLEADO"

    frecuencias = calc_frecuencies(string)
    # pprint(frecuencias)
    # s = sorted(frecuencias.items(), key=lambda item: item[1], reverse=True)
    # pprint(s)
    probabilidades = calc_probabilities(frecuencias)
    # pprint(probabilidades)
    # p = sorted(probabilidades.items(), key=lambda item: item[1], reverse=True)
    # pprint(p)
    # # prob = {"s1": 0.6, "s2": 0.2, "s3": 0.1, "s4": 0.1}
    root = create_tree(probabilidades)
    encode(huff_codes, root)
    print_codes(huff_codes)


if __name__ == "__main__":
    main()

File: Practica/Guia08/src/Huffman.java
import java.util.*;

public class Huffman {
    protected TreeMap<String, Integer> frec = new TreeMap<>();
    protected TreeMap<String, Double> prob = new TreeMap<>();
    protected final TreeMap<String, String> huffcodes = new TreeMap<>();
    protected String msgToEncode;
    protected Nodo root;

    public static String printable(String c) {
        if (c.equals(" ")) {
            return "' '";
        }
        return c;
    }

    public static void main(String[] args) {
        Huffman h = new Huffman();
        String msg = "CALIDO AMANECER SOLEADO";

        h.analizeInput(msg);

        // TreeMap<String, Double> p = new TreeMap<>();
        // p.put("s1", 0.6);
        // p.put("s2", 0.2);
        // p.put("s3", 0.1);
        // p.put("s4", 0.1);
        // h.setProb(p);
        h.generarArbolCodificacion();
        // h.generarCodigos();
        // System.out.println(h);
        // System.out.println("Encoded msg = " + h.getEncodedMsg());
    }

    public void setProb(TreeMap<String, Double> prob) {
        this.prob = prob;
    }

    public void analizeInput(String msg) {
        this.msgToEncode = msg;
        for (int i = 0; i < msg.length(); i++) {
            String c = String.valueOf(msg.charAt(i));
            this.frec.put(c, this.frec.getOrDefault(c, 0) + 1);
        }
        // System.out.println(this.frec);
        this.frec.forEach((k, v) -> this.prob.put(k, (double) v / msg.length()));
        // System.out.println(this.prob);
        // this.frec.forEach((k, v) -> System.out.println(k + " " + v));
        // this.prob.forEach((k, v) -> System.out.println(k + " " + v));
    }

    public void generarArbolCodificacion() {
        PriorityQueue<Nodo> pq = new PriorityQueue<>();
        Nodo hijoIzq, hijoDer, padre;

        for (String s : this.prob.keySet()) {
            pq.add(new Nodo(s, this.prob.get(s), null, null));
        }
        while (pq.size() > 1) {
            hijoIzq = pq.poll();
            hijoDer = pq.poll();
            System.out.println(hijoIzq.simb + " " + hijoIzq.prob + ", " + hijoDer.simb + " " + hijoDer.prob);
            padre = new Nodo(null, hijoIzq.prob + hijoDer.prob, hijoIzq, hijoDer);
            pq.add(padre);
        }
        this.root = pq.poll();
    }

    public void generarCodigos() {
        this.encode(this.root, "");
    }

    private void encode(Nodo root, String s) {
        if (root.isHoja()) {
            this.huffcodes.put(root.simb, s);
        } else {
            encode(root.izq, s + "0");
            encode(root.der, s + "1");
        }
    }

    public String getEncodedMsg() {
        String result = "";
        for (int i = 0; i < this.msgToEncode.length(); i++) {
            String c = String.valueOf(this.msgToEncode.charAt(i));
            result += this.huffcodes.get(c);
        }
        result += "\n";
        return result;
    }

    @Override
    public String toString() {
        StringBuilder huff = new StringBuilder();
        huff.append("Huffman Codes").append("\n");
        for (String s : this.huffcodes.keySet()) {
            huff.append(String.format("word=%s, frec=%d, code=%s\n", Huffman.printable(s), this.frec.get(s),
                    this.huffcodes.get(s)));
        }
        return huff.toString();
    }

    public static class Nodo implements Comparable<Nodo> {
        public final String simb;
        public final double prob;
        public final Nodo izq;
        public final Nodo der;

        public Nodo(String simb, double prob, Nodo izq, Nodo der) {
            this.simb = simb;
            this.prob = prob;
            this.izq = izq;
            this.der = der;
        }

        public boolean isHoja() {
            return this.izq == null && this.der == null;
        }

        @Override
        public int compareTo(Nodo o) {
            return Double.compare(this.prob, o.prob);
        }

        @Override
        public String toString() {
            return "{ simb=" + this.simb + ", prob=" + this.prob + " }";
        }

        public double getProb() {
            return prob;
        }
    }
}